### DISCUSSION TITLE
Displaying a chat log for messages

### DISCUSSION THREAD

**Discussion by ranile** [openmls/openmls#1423] [Q&A]
In a messages application, messages that are sent/received can be viewed after sending/more than once. How would I build such a thing with OpenMLS? I would also like to support multiple devices (e.g. a user may send/receive messages from their desktop computer and also view them on their phone)

 I do not want to store anything in plain text.

**Comment by kkohbrok** [comment]
Hi @hamza1311! 

Regarding message storage in plain text: MLS is geared towards forward secrecy, so key material is deleted as soon as it's been used. That means that you won't be able to decrypt messages twice, even if they are from other clients. In general, MLS mandates a pretty aggressive key deletion schedule, so it's not a good fit for at-rest encryption. My advice would thus be to use separate protocols/solutions for encryption at-rest vs. encryption of message on the wire.

Regarding multi-client scenarios: MLS doesn't have a notion of "users", only clients. So any such notion is going to have to be built on top by the application. The simplest solution would probably be to represent each client of a given user as a separate MLS client and then associate them with the same user as part of whatever identity/authentication scheme you are using on top of MLS.

**Reply by ranile** [reply]
I could share one "client" across devices (otherwise, each one would have to be added into a "room" so they can decrypt the messages, no?)

The major problem I'm facing is encrypting everything for at-rest storage. I want users to be able to access their messages from any device so the encrypted messages would need to be stored on the server. However, that means I must store every message for every user individually. A room with 20 members would need to have every message stored 20 times in the database. That is obviously not ideal. I don't have a solution for it. Any advice would be welcome

**Reply by kkohbrok** [reply]
I still don't quite understand what your scenario is. You want users to be able to create new clients on the fly, join a group that they already have (had?) a client in and immediately get access to all of the group's messages? The protocol will not support that, as that would specifically violate the PCS guarantees it tries to provide (not knowing anything about users). So you'll definitely have to build something around that. One way to go might be to encrypt the messages under a key derived from an MLS exporter key. That would allow all (existing) group members to derive the same key and decrypt the messages. Then maybe upload that key once for each individual user, where each time the key is wrapped with some user-level secret. How you generate/derive that secret leads us to the next problem, where, as the user creates a new client on the fly, it won't have any high-entropy secrets lying around. At that point, the design space it pretty large and really depends on your security requirements, the UX you want for your users and your overall setup. Allowing users to recover secret data essentially from nothing (or just their password) is generally a pretty hard task if you're not prepared to trust the server to some degree.
